Destroy material when a mesh is deleted

ahcorde opened the pull request:
# 🎉 Enhancement 

## Summary

This PR allows to remove materials when a mesh is removed.

There are two cases:
 - When the material is loaded, the method in charge to load the meshes will create a cache of material. I added a method to remove this cache
 - When a mesh is removed the material/s are not removed from the memory. This PR also includes this step.

I had to break the ABI, I'm open to suggestions to be able to backport this PR

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

iche033 commented on ogre2/src/Ogre2MeshFactory.cc:
```diff
@@ -88,6 +89,17 @@ void Ogre2MeshFactory::Clear()
   this->ogreMeshes.clear();
 }
 
+void Ogre2MeshFactory::ClearMaterialsCache()
```
I verified that this fixes the memory leak. A different issue is created after removing these materials though. Here are the steps to reproduce:
1. Insert a model like this one (I just dragged and dropped) https://app.ignitionrobotics.org/GoogleResearch/fuel/models/Weisshai_Great_White_Shark
    * it needs to be a model that references a texture in the dae or obj file but does not have a <material> sdf element
 1. Insert the same model again -> model appear grey 

After debugging a little, I remembered that these materials were created as "templates". Later on in this line https://github.com/ignitionrobotics/ign-rendering/blob/6fd0c03ce32568f8bc91c1041fc169cc7622248f/ogre2/src/Ogre2MeshFactory.cc#L601 a "new instance" is created and assigned to the submesh. This is done by looking for the name of the template material stored in the submesh, getting the template material from the scene, then assigning a copy to the submesh. Now that this template material is removed, subsequent meshes requiring the same material will not be able to find it. 

One possible solution is to remove the code for creating these template materials, i.e. the code here: https://github.com/ignitionrobotics/ign-rendering/blob/6fd0c03ce32568f8bc91c1041fc169cc7622248f/ogre2/src/Ogre2MeshFactory.cc#L455-L471 

we just then need to always create a new material and assign it to the submesh, i.e. replace the logic here: https://github.com/ignitionrobotics/ign-rendering/blob/6fd0c03ce32568f8bc91c1041fc169cc7622248f/ogre2/src/Ogre2MeshFactory.cc#L590-L602


I ran out of time today and haven't found a way to do his cleanly yet..

ahcorde replied:
@iche033, I don't fully understand why this is happening, I would expect this three things

 - when we load a new mesh then the "template" material is loaded and removed.
 - When we remove the mesh we remove the material created fom the template
 - At this point we should not have any material associated with the mesh in memory.
 - Finally we load again the same mesh and I would expect to reload everything again (template and material created from the template).

am I missing something ?

iche033 replied:
> Finally we load again the same mesh and I would expect to reload everything again (template and material created from the template).

we have this check here to skip loading the mesh if it's already loaded in ogre: https://github.com/ignitionrobotics/ign-rendering/blob/6fd0c03ce32568f8bc91c1041fc169cc7622248f/ogre2/src/Ogre2MeshFactory.cc#L167

That was mainly done for performance reason, e.g. when launching a world with hundreds of the same model, esp if they have large complex meshes.

ahcorde commented:
With the last commit, I fixed the issue that @iche033 mentioned and also fixed the small leak

![Selección_098](https://user-images.githubusercontent.com/1933907/118881001-d5976f80-b8f2-11eb-8591-ca1798b22db0.png)

![Selección_099](https://user-images.githubusercontent.com/1933907/118881034-e21bc800-b8f2-11eb-805a-fb2fcb0ccaf2.png)

ahcorde commented:
There is still an issue when you insert more than one model of the same type, from the second model they will not get the material

iche033 commented on ogre/src/OgreMesh.cc:
```diff
@@ -58,6 +58,8 @@ void OgreMesh::Destroy()
 
   auto ogreScene = std::dynamic_pointer_cast<OgreScene>(this->Scene());
 
+  Ogre::MeshManager::getSingleton().remove(
```
this removes the template mesh. Should we have a similar logic as what you did for textures and check reference count to the mesh first before removing it? Same for the `remove` call in `OgreSubMesh::Destroy()`

iche033 commented on ogre/src/OgreMaterial.cc:
```diff
@@ -44,21 +44,52 @@ void OgreMaterial::Destroy()
+        if (this->textureName == res->getName() &&
+          res->getName().find("scene::RenderTexture") == std::string::npos)
+        {
+          this->Scene()->ClearMaterialsCache(this->textureName);
```
Here's an idea to prevent breaking ABI. We can add a new `OgreScene::MeshFactory()` accessor function to `OgreScene` class (and also `Ogre2Scene`) class to get the mesh factory then here we just need to do:

```
OgreScenePtr s = std::dynamic_pointer_cast<OgreScene>(this->Scene());
auto factory = s->MeshFactory();
factory->ClearMaterialCache(this->textureName);
```
This prevents adding and overriding a virtual function in Scene.

iche033 reviewed (commented):
confirms that the latest changes fix the missing texture issue for subsequent models. Memory consumption for repeatedly adding and removing models also look to be stable.

iche033 commented:
> Memory consumption for repeatedly adding and removing models also look to be stable.

Actually that was testing with ogre 1.x. I just tried ogre 2.x, I think there is still some leak.

ahcorde commented:
@mjcarroll and @iche033 I open this PR for review

iche033 commented on ogre/include/ignition/rendering/ogre/OgreMeshFactory.hh:
```diff
@@ -56,7 +56,12 @@ namespace ignition
 
       protected: virtual bool Validate(const MeshDescriptor &_desc);
 
+      /// \brief Remove internal material cache fot a specific material
```
`fot` -> `for`

iche033 commented on ogre/include/ignition/rendering/ogre/OgreMeshFactory.hh:
```diff
@@ -56,7 +56,12 @@ namespace ignition
+
       protected: OgreScenePtr scene;
+
+      protected: std::vector<MaterialPtr> materialCache;
```
add doxygen

iche033 commented on ogre/include/ignition/rendering/ogre/OgreScene.hh:
```diff
@@ -200,6 +200,8 @@ namespace ignition
 
       protected: virtual MaterialMapPtr Materials() const override;
 
+      public: void ClearMaterialsCache(const std::string &_name);
```
add doc

iche033 commented on ogre/src/OgreMeshFactory.cc:
```diff
@@ -106,6 +106,24 @@ bool OgreMeshFactory::IsLoaded(const MeshDescriptor &_desc)
+    }
+    ++it;
+  }
+  this->materialCache.erase(it);
```
add check:

```
if (it != this->dataPtr->materialCache.end())
```

iche033 commented on ogre2/include/ignition/rendering/ogre2/Ogre2Mesh.hh:
```diff
@@ -118,15 +119,21 @@ namespace ignition
+      // Documentation inherited
+      public: virtual void Destroy() override;
+
+      // Documentation inherited
```
update doc

iche033 commented on ogre2/include/ignition/rendering/ogre2/Ogre2MeshFactory.hh:
```diff
@@ -94,6 +94,9 @@ namespace ignition
       /// \brief Pointer to the scene object
       protected: Ogre2ScenePtr scene;
 
+      /// \brief Remove internal material cache fot a specific material
```
`fot` -> `for`

iche033 commented on ogre2/include/ignition/rendering/ogre2/Ogre2Scene.hh:
```diff
@@ -298,6 +298,9 @@ namespace ignition
       /// \brief Create the vaiours storage objects
       private: void CreateStores();
 
+      // Documentation inherited
```
update doc

iche033 commented on ogre2/src/Ogre2Material.cc:
```diff
@@ -44,6 +45,7 @@
 /// \brief Private data for the Ogre2Material class
 class ignition::rendering::Ogre2MaterialPrivate
 {
+  public: std::string hashName;
```
add doc

iche033 commented on ogre2/src/Ogre2Material.cc:
```diff
@@ -87,6 +89,38 @@ void Ogre2Material::Destroy()
+    if (i->second.useCount() == 5)
+    {
+      if (this->dataPtr->hashName == res->getName() &&
+          res->getName().find("scene::RenderTexture") == std::string::npos)
```
`scene::RenderTexture` -> the prefix `scene` is the name of the scene so how about changing this to `scene->Name() + "::RenderTexture"`

iche033 commented on ogre2/src/Ogre2Material.cc:
```diff
@@ -680,6 +715,7 @@ void Ogre2Material::FillUnlitDatablock(Ogre::HlmsUnlitDatablock *_datablock)
         this->ogreDatablock->suggestMapTypeBasedOnTextureType(
         Ogre::PBSM_DIFFUSE));
     _datablock->setTexture(0, texLocation.xIdx, texLocation.texture);
+    std::cerr << "NAME LAEX " << texLocation.texture->getName() << '\n';
```
remove line

iche033 commented on ogre2/src/Ogre2Mesh.cc:
```diff
@@ -280,6 +292,39 @@ Ogre2SubMesh::~Ogre2SubMesh()
   this->Destroy();
 }
 
+void Ogre2SubMesh::SetMeshName(const std::string &_name)
```
add `/////` separation line

iche033 commented on ogre2/src/Ogre2Mesh.cc:
```diff
@@ -280,6 +292,39 @@ Ogre2SubMesh::~Ogre2SubMesh()
+  this->dataPtr->subMeshName = _name;
+}
+
+void Ogre2SubMesh::Destroy()
```
add `/////` separation line

iche033 commented on ogre2/src/Ogre2Mesh.cc:
```diff
@@ -280,6 +292,39 @@ Ogre2SubMesh::~Ogre2SubMesh()
+    auto iend = meshManager->getResourceIterator().end();
+    for (auto i = meshManager->getResourceIterator().begin(); i != iend;)
+    {
+      // A use count of 4 means that only RGM, RM and MeshManager have
```
use count of 3?

iche033 commented on ogre2/src/Ogre2MeshFactory.cc:
```diff
@@ -58,6 +58,7 @@
 /// \brief Private data for the Ogre2MeshFactory class
 class ignition::rendering::Ogre2MeshFactoryPrivate
 {
+  public: std::vector<MaterialPtr> materialCache;
```
add doc

iche033 commented on ogre2/src/Ogre2MeshFactory.cc:
```diff
@@ -88,6 +89,24 @@ void Ogre2MeshFactory::Clear()
   this->ogreMeshes.clear();
 }
 
+void Ogre2MeshFactory::ClearMaterialsCache(const std::string &_name)
```
add `/////` separation line

iche033 commented on ogre/src/OgreMaterial.cc:
```diff
@@ -44,21 +44,53 @@ void OgreMaterial::Destroy()
+
     matManager.remove(this->ogreMaterial->getName());
     this->ogreMaterial.setNull();
-  }
```
I think this is removed by accident. 

Windows build caught a compile error:
C:\Jenkins\workspace\ign_rendering-pr-win\ws\ign-rendering\ogre\src\OgreMaterial.cc(94,1): error C2059: syntax error: '}'

iche033 reviewed (commented):
nice! the latest changes fix the ogre2 memory leak for me. I just have a few coding style and doc comments

iche033 commented on ogre2/src/Ogre2Mesh.cc:
```diff
@@ -280,6 +292,40 @@ Ogre2SubMesh::~Ogre2SubMesh()
+    auto iend = meshManager->getResourceIterator().end();
+    for (auto i = meshManager->getResourceIterator().begin(); i != iend;)
+    {
+      // A use count of 43 means that only RGM and RM have
```
`43` -> `3`

iche033 reviewed (approved):
looks good to me. just one minor typo.
